### PR TITLE
[codegen/go] Register resource modules/packages

### LINKED ISSUE #5801: [codegen/go] Register package/module resources
Go side of https://github.com/pulumi/pulumi/pull/5806

The generated code will need to call [`RegisterResourcePackage`/`RegisterResourceModule`](https://github.com/pulumi/pulumi/blob/260620430c4c8a41a8d7adfbd66f68ba6fd1e168/sdk/go/pulumi/rpc.go#L682-L718) in Go package `init` functions.

Looks like we'll also need to tweak the signatures of `RegisterResourcePackage`, `RegisterResourceModule`, and the interfaces to match similar tweaks we've made to [Python](https://github.com/pulumi/pulumi/pull/5805) and [Node.js](https://github.com/pulumi/pulumi/pull/5837).

### PR DESCRIPTION
**Pull request by lblackstone**
Fix #5801

### CODE REVIEW

File: pkg/codegen/go/gen.go [hunk 1]
```diff
@@ -1323,6 +1332,102 @@ func (pkg *pkgContext) genConfig(w io.Writer, variables []*schema.Property) erro
...
+		fmt.Fprintf(w, "}\n")
+	}
+
+	if provider != nil {
+		imports := codegen.NewStringSet(
+			"github.com/blang/semver",
+			"github.com/pulumi/pulumi/sdk/v2/go/pulumi",
+		)
+
+		pkg.genHeader(w, []string{"fmt"}, imports)
```
Comment (justinvp) [inline pkg/codegen/go/gen.go:1378]: What does the generated code look like for a provider that doesn't have any sub-modules (like `random`)? I would expect the same file to have calls to both `RegisterResourcePackage` and `RegisterResourceModule`. Like https://github.com/pulumi/pulumi-random/blob/master/sdk/python/pulumi_random/__init__.py

**Review by justinvp — COMMENTED** [review]
Have you tried generating any providers locally that you could share in a branch/commit (to see the diff)? I'd love to see what `aws`, `random`, and `kubernetes` look like.

Also, what do you think about adding some codegen tests, like we did for Python?
- https://github.com/pulumi/pulumi/blob/aec642e8c9fc2e78365d84850a8a4e4676424cb5/pkg/codegen/internal/test/testdata/simple-enum-schema/python/pulumi_plant_provider/tree/v1/__init__.py#L31
- https://github.com/pulumi/pulumi/blob/aec642e8c9fc2e78365d84850a8a4e4676424cb5/pkg/codegen/internal/test/testdata/simple-enum-schema/python/pulumi_plant_provider/__init__.py#L36

LGTM otherwise.

**Comment by lblackstone** [comment]
> Have you tried generating any providers locally that you could share in a branch/commit (to see the diff)? I'd love to see what aws, random, and kubernetes look like.

You can see the result in the `init.go` files for the downstream codegen tests. e.g. https://github.com/pulumi-bot/pulumi-azuread/compare/d5c6029..310753e#diff-3eb4898547e678835170ac6c0a4c6289624b30771b535d22e90438cf04f05573

File: pkg/codegen/go/gen.go (hunk 1 above)
Comment (lblackstone) [reply pkg/codegen/go/gen.go:1378]: Yeah, it looks like I don't have this quite right yet. https://github.com/pulumi-bot/pulumi-random/compare/f7e21f1..c596e19#diff-572719db95a5ff9f62d8b4d256f04462e146ae584e1d4612f39702aac242e4b3

**Comment by lblackstone** [comment]
> Also, what do you think about adding some codegen tests, like we did for Python?

Will do

**Comment by justinvp** [comment]
> You can see the result in the `init.go` files for the downstream codegen tests.

Yep, I see that now. I thought those weren't working (due to not being up-to-date) when I first looked at the PR.

File: pkg/codegen/go/gen.go [hunk 2]
```diff
@@ -1323,6 +1332,86 @@ func (pkg *pkgContext) genConfig(w io.Writer, variables []*schema.Property) erro
...
+		fmt.Fprintf(w, "\tif err != nil {\n")
+		fmt.Fprintf(w, "\t\tpanic(fmt.Errorf(\"failed to determine package version: %%v\", err))\n")
+		fmt.Fprintf(w, "\t}\n")
+		fmt.Fprintf(w, "\tpulumi.RegisterResourcePackage(\n")
+		fmt.Fprintf(w, "\t\t%q,\n", pkg.pkg.Name)
+		fmt.Fprintf(w, "\t\t&module{version},\n")
+		fmt.Fprintf(w, "\t)\n")
+		fmt.Fprintf(w, "}\n")
+	} else {
+		imports.Add("github.com/pulumi/pulumi-kubernetes/sdk/v2/go/kubernetes")
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go:1377]: Is this import of the k8s SDK intentional?

Comment (lblackstone) [reply pkg/codegen/go/gen.go]: Nope - it's fixed in the updates I just pushed

**Comment by lblackstone** [comment]
Hmm, for some reason the `ImportBasePath` isn't being set on the goInfo for GCP, which results in invalid generated code for that provider.

File: pkg/codegen/go/gen.go [hunk 3]
```diff
@@ -1710,14 +1823,14 @@ func getEnvOrDefault(def interface{}, parser envParser, vars ...string) interfac
 func PkgVersion() (semver.Version, error) {
 	type sentinal struct{}
 	pkgPath := reflect.TypeOf(sentinal{}).PkgPath()
-	re := regexp.MustCompile("^.*/pulumi-%s/sdk/v(\\d+)*")
+	re := regexp.MustCompile("^.*/pulumi-%s/sdk/(v\\d+)*")
```
Comment (pgavlin) [inline pkg/codegen/go/gen.go:1826]:
```suggestion
	re := regexp.MustCompile("^.*/pulumi-%s/sdk(/v\\d+)*")
```

Comment (pgavlin) [inline pkg/codegen/go/gen.go:1826]:
```suggestion
	re := regexp.MustCompile("^.*/pulumi-%s/sdk(/v\\d+)?")
```

**Review by justinvp — APPROVED** [review]
LGTM

**Pull request merged**